homeassistant/components/sensor/time_date.py

```diff
@@ -79,10 +87,21 @@ def icon(self):
         else:
             return 'mdi:clock'
 
-    @asyncio.coroutine
-    def async_update(self):
-        """Get the latest data and updates the states."""
-        time_date = dt_util.utcnow()
+    def get_next_interval(self, now=None):
+        """Compute next time an update should occur."""
+        if self.type == 'date':
+            interval = 86400
+        elif self.type == 'beat':
+            interval = 86.4
+        else:
+            interval = 60
+        if now is None:
+            now = dt_util.utcnow()
+        next_timestamp = int(dt_util.as_timestamp(now))
```

Comment:
You're right, I didn't fully copy your algorithm. I was merely trying to point out that Python has a built-in class to work with time periods.